File: jas_ocaml/test/algorithms/simplify_test.ml
(** Polyline-to-Bezier simplification tests.
    Mirrors jas_dioxus/src/algorithms/simplify.rs tests. *)

open Jas.Simplify
module E = Jas.Element

let is_move = function E.MoveTo _ -> true | _ -> false
let is_line = function E.LineTo _ -> true | _ -> false
let is_curve = function E.CurveTo _ -> true | _ -> false
let is_close = function E.ClosePath -> true | _ -> false

let count p lst = List.length (List.filter p lst)
let last lst = List.nth lst (List.length lst - 1)

let () =
  Alcotest.run "Simplify" [
    "simplify_polyline", [
      Alcotest.test_case "empty input returns empty" `Quick (fun () ->
        assert (simplify_polyline [] 0.5 true = []));

      Alcotest.test_case "two points emits MoveTo + LineTo" `Quick (fun () ->
        let out = simplify_polyline [(0.0, 0.0); (10.0, 0.0)] 0.5 false in
        assert (List.length out = 2);
        assert (is_move (List.nth out 0));
        assert (is_line (List.nth out 1)));

      Alcotest.test_case "two points closed appends ClosePath" `Quick (fun () ->
        let out = simplify_polyline [(0.0, 0.0); (10.0, 0.0)] 0.5 true in
        assert (List.length out = 3);
        assert (is_move (List.nth out 0));
        assert (is_line (List.nth out 1));
        assert (is_close (List.nth out 2)));

      Alcotest.test_case "square keeps lines (no curves)" `Quick (fun () ->
        (* Closed square — every edge is straight, so the output should
           be 4 LineTo + ClosePath after the initial MoveTo. *)
        let sq = [(0.0, 0.0); (10.0, 0.0); (10.0, 10.0); (0.0, 10.0)] in
        let out = simplify_polyline sq 0.5 true in
        assert (count is_curve out = 0);
        assert (count is_line out = 4);
        assert (is_close (last out)));

      Alcotest.test_case "circle sampling recovers curves" `Quick (fun () ->
        let n = 32 in
        let r = 50.0 in
        let pts = List.init n (fun i ->
          let t = 2.0 *. Float.pi *. float_of_int i /. float_of_int n in
          (r *. cos t, r *. sin t)) in
        let out = simplify_polyline pts 0.5 true in
        assert (count is_curve out > 0);
        assert (count is_line out = 0);
        assert (is_close (last out)));
    ];

    "detect_corners", [
      Alcotest.test_case "closed square -> all 4 vertices" `Quick (fun () ->
        let sq = [(0.0, 0.0); (10.0, 0.0); (10.0, 10.0); (0.0, 10.0)] in
        let corners = detect_corners sq default_corner_angle true in
        assert (corners = [0; 1; 2; 3]));

      Alcotest.test_case "collinear points -> none" `Quick (fun () ->
        let line = List.init 10 (fun i -> (float_of_int i, 0.0)) in
        let corners = detect_corners line default_corner_angle false in
        assert (corners = []));

      Alcotest.test_case "25 degree turn below threshold -> none" `Quick (fun () ->
        let angle = 25.0 *. Float.pi /. 180.0 in
        let pts = [
          (0.0, 0.0);
          (10.0, 0.0);
          (10.0 +. 10.0 *. cos angle, 10.0 *. sin angle);
        ] in
        let corners = detect_corners pts default_corner_angle false in
        assert (corners = []));

      Alcotest.test_case "45 degree turn above threshold -> corner" `Quick (fun () ->
        let angle = 45.0 *. Float.pi /. 180.0 in
        let pts = [
          (0.0, 0.0);
          (10.0, 0.0);
          (10.0 +. 10.0 *. cos angle, 10.0 *. sin angle);
        ] in
        let corners = detect_corners pts default_corner_angle false in
        assert (corners = [1]));

      Alcotest.test_case "open polyline endpoints are not corners" `Quick (fun () ->
        let pts = [(0.0, 0.0); (5.0, 0.0); (10.0, 0.0)] in
        let corners = detect_corners pts default_corner_angle false in
        assert (corners = []));
    ];
  ]

File: jas/algorithms/simplify_test.py
"""Polyline simplification tests.

Mirrors the ``#[cfg(test)]`` module in
``jas_dioxus/src/algorithms/simplify.rs`` for cross-language behavioral
equivalence.
"""

from __future__ import annotations

import math

from algorithms.simplify import (
    DEFAULT_CORNER_ANGLE,
    detect_corners,
    simplify_polyline,
)
from geometry.element import ClosePath, CurveTo, LineTo, MoveTo


def test_empty_input_returns_empty():
    assert simplify_polyline([], 0.5, True) == []


def test_two_points_emits_lineto():
    out = simplify_polyline([(0.0, 0.0), (10.0, 0.0)], 0.5, False)
    assert len(out) == 2
    assert isinstance(out[0], MoveTo)
    assert isinstance(out[1], LineTo)


def test_detect_corners_on_square():
    # Closed unit square -- every vertex is a 90 degree corner.
    sq = [(0.0, 0.0), (10.0, 0.0), (10.0, 10.0), (0.0, 10.0)]
    corners = detect_corners(sq, DEFAULT_CORNER_ANGLE, True)
    assert corners == [0, 1, 2, 3]


def test_detect_corners_on_collinear_points():
    # Collinear points should not yield corners.
    line = [(float(i), 0.0) for i in range(10)]
    corners = detect_corners(line, DEFAULT_CORNER_ANGLE, False)
    assert corners == [], f"got unexpected corners on a straight line: {corners}"


def test_detect_corners_below_threshold_is_smooth():
    # 25-degree turn -- below the 30-degree threshold, no corner.
    angle = math.radians(25.0)
    pts = [
        (0.0, 0.0),
        (10.0, 0.0),
        (10.0 + 10.0 * math.cos(angle), 10.0 * math.sin(angle)),
    ]
    corners = detect_corners(pts, DEFAULT_CORNER_ANGLE, False)
    assert corners == [], f"25 degree turn should not be a corner, got {corners}"


def test_detect_corners_above_threshold_is_corner():
    # 45-degree turn -- above the 30-degree threshold, marked.
    angle = math.radians(45.0)
    pts = [
        (0.0, 0.0),
        (10.0, 0.0),
        (10.0 + 10.0 * math.cos(angle), 10.0 * math.sin(angle)),
    ]
    corners = detect_corners(pts, DEFAULT_CORNER_ANGLE, False)
    assert corners == [1]


def test_simplify_square_keeps_lines():
    # Closed square -- every edge is straight, so the output should be
    # 4 LineTo + ClosePath after the initial MoveTo. No CurveTo.
    sq = [(0.0, 0.0), (10.0, 0.0), (10.0, 10.0), (0.0, 10.0)]
    out = simplify_polyline(sq, 0.5, True)
    curve_count = sum(1 for c in out if isinstance(c, CurveTo))
    line_count = sum(1 for c in out if isinstance(c, LineTo))
    assert curve_count == 0, "square should fit with no curves"
    assert line_count == 4, "square should fit as 4 LineTo segments"
    assert isinstance(out[-1], ClosePath)


def test_simplify_circle_recovers_curves():
    # 32-segment regular circle sampling -- should fit as curves with no
    # corners and no LineTo.
    n = 32
    r = 50.0
    pts = [
        (r * math.cos(2.0 * math.pi * i / n), r * math.sin(2.0 * math.pi * i / n))
        for i in range(n)
    ]
    out = simplify_polyline(pts, 0.5, True)
    curve_count = sum(1 for c in out if isinstance(c, CurveTo))
    line_count = sum(1 for c in out if isinstance(c, LineTo))
    assert curve_count > 0, "circle sampling should fit at least one CurveTo"
    assert line_count == 0, "circle sampling should not produce LineTo"
    assert isinstance(out[-1], ClosePath)


def test_open_polyline_endpoints_are_not_corners():
    # Three collinear points -- endpoints at index 0 and 2 must not be
    # reported as corners, only vertex 1 could (and it shouldn't here
    # because it is collinear).
    pts = [(0.0, 0.0), (5.0, 0.0), (10.0, 0.0)]
    corners = detect_corners(pts, DEFAULT_CORNER_ANGLE, False)
    assert corners == [], f"got {corners}"
